infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/secret/FileSecretApplier.java

```diff
@@ -0,0 +1,139 @@
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment.PodRole;
+
+/**
+ * Mounts Kubernetes secret with specific annotations as an file in workspace containers. Via
+ * devfile, allows per-component control of secret applying and path overrides using specific
+ * property.
+ */
+@Beta
+@Singleton
+public class FileSecretApplier extends KubernetesSecretApplier<KubernetesEnvironment> {
+
+  static final String ANNOTATION_MOUNT_PATH = ANNOTATION_PREFIX + "/" + "mount-path";
+
+  /**
+   * Applies secret as file into workspace containers, respecting automount attribute and optional
+   * devfile automount property and/or mount path override.
+   *
+   * @param env kubernetes environment with workspace containers configuration
+   * @param secret source secret to apply
+   * @throws InfrastructureException on misconfigured secrets or other apply error
+   */
+  @Override
+  public void applySecret(KubernetesEnvironment env, Secret secret) throws InfrastructureException {
+    String mountPath = secret.getMetadata().getAnnotations().get(ANNOTATION_MOUNT_PATH);
+    boolean secretAutomount =
+        Boolean.parseBoolean(secret.getMetadata().getAnnotations().get(ANNOTATION_AUTOMOUNT));
+    if (mountPath == null) {
+      throw new InfrastructureException(
+          format(
+              "Unable to mount secret '%s': It is configured to be mounted as a file but the mount path was not specified. Please define the '%s' annotation on the secret to specify it.",
+              secret.getMetadata().getName(), ANNOTATION_MOUNT_PATH));
+    }
+
+    Volume volumeFromSecret =
+        new VolumeBuilder()
+            .withName(secret.getMetadata().getName())
+            .withSecret(
+                new SecretVolumeSourceBuilder()
+                    .withNewSecretName(secret.getMetadata().getName())
+                    .build())
+            .build();
+
+    for (PodData podData : env.getPodsData().values()) {
+      if (!podData.getRole().equals(PodRole.DEPLOYMENT)) {
+        continue;
+      }
+      if (podData
+          .getSpec()
+          .getVolumes()
+          .stream()
+          .anyMatch(v -> v.getName().equals(volumeFromSecret.getName()))) {
+        volumeFromSecret.setName(volumeFromSecret.getName() + "_" + NameGenerator.generate("", 6));
+      }
+
+      podData.getSpec().getVolumes().add(volumeFromSecret);
+
+      for (Container container : podData.getSpec().getContainers()) {
+        Optional<ComponentImpl> component = getComponent(env, container.getName());
+        // skip components that explicitly disable automount
+        if (component.isPresent() && isComponentAutomountFalse(component.get())) {
+          continue;
+        }
+        // if automount disabled globally and not overridden in component
+        if (!secretAutomount
+            && (component.isEmpty() || !isComponentAutomountTrue(component.get()))) {
```

Comment:
`!isComponentAutomountTrue(component.get())` isn't it same as `isComponentAutomountFalse(component.get())` ?

Reply:
not exactly. they both return `true` only when property set explicitly. So when property not set, `isComponentAutomountFalse()` will return false,  when set to `false` it will return true. And here we need to check it is not set or ot is not true. So having different methods more strightforward.

Reply:
ouch, that's tricky. Can you please javadoc those methods

Reply:
ok, I see you've already document it. thanks